deltacat/aws/clients.py

```diff
@@ -1,22 +1,52 @@
 import logging
 from functools import lru_cache
+from requests import Session
 from typing import Optional
+from http import HTTPStatus
 
 import boto3
 from boto3.exceptions import ResourceNotExistsError
 from boto3.resources.base import ServiceResource
 from botocore.client import BaseClient
 from botocore.config import Config
+from requests.adapters import HTTPAdapter, Retry, Response
 
 from deltacat import logs
 from deltacat.aws.constants import BOTO_MAX_RETRIES
 
 logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
 
 BOTO3_PROFILE_NAME_KWARG_KEY = "boto3_profile_name"
+INSTANCE_METADATA_SERVICE_IPV4_URI = "http://169.254.169.254/latest/meta-data/"  # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+
+
+def block_until_instance_metadata_service_returns_success(
+    total_number_of_retries=10,
+    backoff_factor=1,
+) -> Optional[Response]:
+    # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+    with Session() as session:
+        retries = Retry(
+            total=total_number_of_retries,
+            backoff_factor=backoff_factor,  # A backoff factor to apply between attempts after the second try: {backoff factor} * (2 ** ({number of previous retries}))
```

Comment:
Shouldn't we have equal spaced retries? For example, say the we retried 5 times, next we have to wait atleast 256s to retry again. We can retry every 1s for atleast 10 mins before failing. Generally backoff is desired in case of downstream service throttles where we need to give some time for the downstream to recover. In this case, it's the startup issue.

Reply:
Revised using the tenacity library. Policy set to retrying every 2 seconds and stopping 10 minutes before failing